presto-main/src/main/java/com/facebook/presto/sql/planner/iterative/rule/DereferencePushDownFilter.java

```diff
@@ -0,0 +1,89 @@
+import com.facebook.presto.sql.planner.plan.Assignments;
+import com.facebook.presto.sql.planner.plan.FilterNode;
+import com.facebook.presto.sql.planner.plan.PlanNode;
+import com.facebook.presto.sql.planner.plan.ProjectNode;
+import com.facebook.presto.sql.tree.DereferenceExpression;
+import com.facebook.presto.sql.tree.Expression;
+import com.facebook.presto.sql.tree.ExpressionTreeRewriter;
+import com.google.common.collect.ImmutableMap;
+
+import java.util.List;
+import java.util.Map;
+import java.util.function.Function;
+
+import static com.facebook.presto.sql.planner.SymbolsExtractor.extractAll;
+import static com.facebook.presto.sql.planner.plan.Patterns.filter;
+import static com.google.common.collect.Iterables.getOnlyElement;
+import static java.util.stream.Collectors.toList;
+import static java.util.stream.Collectors.toMap;
+
+public class DereferencePushDownFilter
+        extends DereferencePushDownRule<FilterNode>
+{
+    public DereferencePushDownFilter(Metadata metadata, SqlParser sqlParser)
+    {
+        super(metadata, sqlParser, filter());
+    }
+
+    @Override
+    protected PlanNode dereferencePushDown(Context context, FilterNode filterNode, Map<DereferenceExpression, Symbol> expressions, Assignments assignments)
+    {
+        List<Symbol> symbols = filterNode.getOutputSymbols().stream().collect(toList());
+        Map<Expression, Symbol> pushdownExpressions = expressions.entrySet().stream()
+                .filter(entry -> !isCastRowType(entry.getKey()))
+                .filter(entry -> symbols.contains(getOnlyElement(extractAll(entry.getKey()))))
+                .collect(toMap(Map.Entry::getKey, Map.Entry::getValue));
+
+        if (pushdownExpressions.isEmpty()) {
+            return filterNode;
+        }
+
+        Map<Symbol, Expression> pushdownDereferences = pushdownExpressions.entrySet().stream().collect(toMap(Map.Entry::getValue, Map.Entry::getKey));
+        List<DereferenceExpression> predicates = extractDereference(filterNode.getPredicate());
+        Map<Symbol, DereferenceExpression> predicateDereferences = predicates.stream()
+                .distinct()
+                .filter(expression -> !expressions.containsKey(expression))
+                .collect(toMap(expression -> getSymbol(expression, context, metadata, sqlParser), Function.identity()));
+
+        Map<DereferenceExpression, Symbol> predicateExpressions = predicateDereferences.entrySet().stream()
+                .collect(toMap(Map.Entry::getValue, Map.Entry::getKey));
+
+        List<Symbol> outputs = filterNode.getOutputSymbols().stream()
+                .filter(symbol -> assignments.getMap().containsKey(symbol))
+                .collect(toList());
+
+        Assignments.Builder pushdownBuilder = Assignments.builder();
+        pushdownBuilder.putAll(pushdownDereferences).putIdentities(outputs);
+        for (Map.Entry<Symbol, DereferenceExpression> entry : predicateDereferences.entrySet()) {
```

Comment:
perhaps, simplify to

```
        Assignments newAssignments = Assignments.builder()
                .putAll(pushdownDereferences)
                .putIdentities(outputs)
                .putAll(predicateDereferences)
                .build();
        ProjectNode child = new ProjectNode(context.getIdAllocator().getNextId(), filterNode.getSource(), newAssignments);
```

this would require changing Assignments a little:

```
        public Builder putAll(Map<Symbol, ? extends Expression> assignments)
        {
            for (Entry<Symbol, ? extends Expression> assignment : assignments.entrySet()) {
                put(assignment.getKey(), assignment.getValue());
            }
            return this;
        }
```